### PR TITLE
feat: message context menu (AR-961) (AR-1176)

### PR DESCRIPTION
**Pull request by trOnk12**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-961" title="AR-961" target="_blank">SubtaskAR-961</a>  Implement Message Contextual Sliding Menu
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

- Create a context menu for the messages
- Copy the message
- make the title in MenuModalSheetItem nullable so that we can create a MenuModalSheetLayout without a title and an icon

### Attachments (Optional)

Screenshot 2022-03-18 at 11 54 46
Screenshot 2022-03-18 at 11 55 02

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreenState.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        editMessage?.messageSource
+    }
+
+    fun showEditContextMenu(message: Message) {
+        editMessage = message
+        coroutineScope.launch { modalBottomSheetState.animateTo(ModalBottomSheetValue.Expanded) }
+    }
+
+    fun copyMessage() {
+        editMessage!!.messageContent.let { messageContent ->
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreenState.kt:67]: Would it be possible to change `!!` to `?`. Here a NPE happening it's a real possibility I think

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt [hunk 2]
```diff
@@ -85,6 +86,7 @@ class ConversationViewModel @Inject constructor(
                         (message.content as? MessageContent.Text)?.value ?: "content is not available"
                     )
                 ),
+                messageSource= MessageSource.CurrentUser,
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:89]: Just a small formatting nitpicking
```suggestion
                messageSource = MessageSource.CurrentUser,
```

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreenState.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import com.wire.android.R
+import com.wire.android.ui.home.conversations.model.Message
+import com.wire.android.ui.home.conversations.model.MessageContent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun rememberConversationScreenState(
+    context: Context = LocalContext.current,
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreenState.kt:29]: I'm not having the full context on the usage of this `rememberConversationScreenState` component or maybe just missing something here, but what/where would be the cases in which we'll be passing a parent context ? Would it make sense to delete this parameter and just declare it inside this composable ?

Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreenState.kt:29]: totally makes sense

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 4]
```diff
@@ -34,35 +40,108 @@ fun ConversationScreen(
...
+                        MenuItemIcon(
+                            id = R.drawable.ic_edit,
+                            contentDescription = stringResource(R.string.content_description_edit_the_message)
+                        )
+                    },
+                    title = stringResource(R.string.label_edit),
+                )
+            }
+        add {
+            CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.error) {
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:123]: I think we can use here the `Tint()` extension function to make it prettier

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 5]
```diff
@@ -34,35 +40,108 @@ fun ConversationScreen(
...
                 )
             }
         )
     }
 }
 
+@Composable
+private fun EditMessageMenuItems(
+    editMessageSource: MessageSource?,
+    onCopyMessage: () -> Unit
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:95]: can we also add a `onMessageDeleted()` lambda?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 6]
```diff
@@ -34,35 +40,108 @@ fun ConversationScreen(
...
+                    }
                 )
             }
         )
     }
 }
 
+@Composable
+private fun EditMessageMenuItems(
+    editMessageSource: MessageSource?,
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:94]: Why is this declared as nullable?

**Review by yamilmedina — APPROVED** [review]
great work, looks good 🥇

**Review by gongracr — APPROVED** [review]
Small comments but looking tasty otherwise 🍙 🥢

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt (hunk 6 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:94]: will add a TODO for now

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt (hunk 5 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:95]: will add a TODO for now

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt (hunk 4 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:123]: will add a TODO for now

**Pull request merged**